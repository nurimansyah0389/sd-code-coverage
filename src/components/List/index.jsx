import React, { Component, useState } from 'react';
import PropTypes from 'prop-types';

const Button = () => <button type="submit">Add Todo</button>
const Form = ({ onSubmit }) => {

    const [todo, setTodo] = useState('');

    const handleChange = e => {
        const { value } = e.target;
        setTodo(value);
    };

    const handleOnSubmit = e => {
        e.preventDefault();
        onSubmit(todo);
        setTodo('');
    };

    return (
        <form onSubmit={handleOnSubmit}>
            <div>
                <label>What will you do next?</label>
                <input type="text" name="todo" id="todo" value={todo} onChange={handleChange}/>
                <Button />
            </div>
        </form>
    );
};
Form.propTypes = {
    onSubmit: PropTypes.func
};

class List extends Component {

    static propTypes = {
        data: PropTypes.array
    };

    state = {
        data: this.props.data || []
    };

    handleOnSubmit = value => {
        const data = this.state.data.slice();
        data.push(value);
        this.setState({ data });
    };

    handleRemove = item => {
        const data = this.state.data.slice().filter(row => row !== item);
        this.setState({ data });
    };

    render = () => {
        const { data } = this.state;

        return (
            <div>
                <ul>
                    {data.map((item, index) => <li key={index}>{item} - <button type="button" onClick={() => this.handleRemove(item)}>Delete</button></li>)}
                </ul>
                <hr/>
                <Form onSubmit={this.handleOnSubmit} />
            </div>
        );
    };
};
export default List;