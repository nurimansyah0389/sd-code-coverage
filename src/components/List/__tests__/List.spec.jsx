import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import List from '../';

describe('List Spec', () => {
    const data = ['First Todo', 'Second Todo', 'Third Todo'];

    it('Should render without error', () => {
        const tree = Renderer.create(<List />);
        expect(tree.toJSON()).toMatchSnapshot();
    });
    it('Should render todo data from the initial data', () => {
        const wrapper = shallow(<List data={data} />);
        const instance = wrapper.instance();

        expect(instance.state.data).toHaveLength(data.length);
    });
    it('Should show add new todo when "Add Todo" button clicked', () => {
        const wrapper = mount(<List data={data} />);
        const instance = wrapper.instance();
        const Form = wrapper.find('Form');
        const Input = Form.find('[name="todo"]');
        const Button = Form.find('Button');
        
        Input.simulate('change', { target: { value: 'Playing footbal' } });
        Button.simulate('submit');

        expect(instance.state.data).toHaveLength(4);
        expect(instance.state.data.includes('Playing footbal')).toBeTruthy();
    });
    it('Should delete selected todo item', () => {
        const wrapper = shallow(<List data={data} />);
        const instance = wrapper.instance();
        const selectedItem = wrapper.find('ul > li').at(0);
        const Button = selectedItem.find('button');

        Button.simulate('click');

        expect(instance.state.data).toHaveLength(2);
        expect(instance.state.data.includes('First Todo')).toBeFalsy();
    });
});