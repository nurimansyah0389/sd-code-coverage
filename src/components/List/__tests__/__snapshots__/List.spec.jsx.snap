// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Spec Should render without error 1`] = `
<div>
  <ul />
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        What will you do next?
      </label>
      <input
        id="todo"
        name="todo"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </div>
  </form>
</div>
`;
